vision/diffusion_mnist/diffusion_mnist.jl

```diff
@@ -0,0 +1,385 @@
+"""
+expand_dims(x::AbstractVecOrMat, dims::Int=2) = reshape(x, (ntuple(i -> 1, dims)..., size(x)...))
+
+"""
+Helper function that reverses the order of dimensions.
+"""
+reverse_dims(x) = permutedims(x, reverse(ntuple(i -> i, length(size(x)))))
+
+"""
+Makes the UNet struct callable and shows an example of a "Functional" API for modeling in Flux. \n
+
+Notes:
+    ```julia
+    @assert size(a) == (28, 28, 1, 32, 32)
+    @assert size(b) == (32, 32)
+    o = reverse_dims(a) .+ reverse_dims(b)' |> reverse_dims
+    size(o)  # (28, 28, 1, 32, 32)
+    ```
+    Is used to broadcast an operation from b to a 
+    without the cost of a copy (using expand_dims for example)
+    which is not allowed (mutating vector) by Zygote in 
+    a custom model.
+"""
+function (unet::UNet)(x, t)
+    # Embedding
+    embed = unet.gaussfourierproj(t)
+    embed = swish.(unet.linear(embed))
+    # Encoder
+    h1 = unet.conv1(x)
+    h1 = reverse_dims(h1) .+ unet.dense1(embed)' |> reverse_dims
+    h1 = unet.gnorm1(h1)
+    h1 = swish.(h1)
+    h2 = unet.conv2(h1)
+    h2 = reverse_dims(h2) .+ unet.dense2(embed)' |> reverse_dims
+    h2 = unet.gnorm2(h2)
+    h2 = swish.(h2)
+    h3 = unet.conv3(h2)
+    h3 = reverse_dims(h3) .+ unet.dense3(embed)' |> reverse_dims
+    h3 = unet.gnorm3(h3)
+    h3 = swish.(h3)
+    h4 = unet.conv4(h3)
+    h4 = reverse_dims(h4) .+ unet.dense4(embed)' |> reverse_dims
+    h4 = unet.gnorm4(h4)
+    h4 = swish.(h4)
+    # Decoder
+    h = unet.tconv4(h4)
+    h = reverse_dims(h) .+ unet.dense5(embed)' |> reverse_dims
+    h = unet.tgnorm4(h)
+    h = swish.(h)
+    h = unet.tconv3(cat(h, h3; dims=3))
+    h = reverse_dims(h) .+ unet.dense6(embed)' |> reverse_dims
+    h = unet.tgnorm3(h)
+    h = swish.(h)
+    h = unet.tconv2(cat(h, h2, dims=3))
+    h = reverse_dims(h) .+ unet.dense7(embed)' |> reverse_dims
+    h = unet.tgnorm2(h)
+    h = swish.(h)
+    h = unet.tconv1(cat(h, h1, dims=3))
+    # Scaling Factor
+    reverse_dims(h) ./ unet.marginal_prob_std(t) |> reverse_dims
+end
+
+"""
+Model loss following the denoising score matching objectives:
+
+# Notes
+Denoising score matching objective:
+```julia
+min wrt. θ (
+    𝔼 wrt. 𝘵 ∼ 𝒰(0, 𝘛)[
+        λ(𝘵) * 𝔼 wrt. 𝘹(0) ∼ 𝒫₀(𝘹) [
+            𝔼 wrt. 𝘹(t) ∼ 𝒫₀ₜ(𝘹(𝘵)|𝘹(0)) [
+                (||𝘚₀(𝘹(𝘵), 𝘵) - ∇ log [𝒫₀ₜ(𝘹(𝘵) | 𝘹(0))] ||₂)²
+            ]
+        ]
+    ]
+)
+``` 
+Where 𝒫₀ₜ(𝘹(𝘵) | 𝘹(0)) and λ(𝘵), are available analytically and
+𝘚₀(𝘹(𝘵), 𝘵) is estimated by a U-Net architecture.
+
+# References:
+http://www.iro.umontreal.ca/~vincentp/Publications/smdae_techreport.pdf \n
+https://yang-song.github.io/blog/2021/score/#estimating-the-reverse-sde-with-score-based-models-and-score-matching \n
+https://yang-song.github.io/blog/2019/ssm/
+"""
+function model_loss(model, x, device, ϵ=1.0f-5)
+    batch_size = size(x)[end]
+    # (batch) of random times to approximate 𝔼[⋅] wrt. 𝘪 ∼ 𝒰(0, 𝘛)
+    random_t = rand(Float32, batch_size) .* (1.0f0 - ϵ) .+ ϵ |> device
```

Comment:
what is this -eps + eps for?

Reply:
[Here](https://colab.research.google.com/drive/120kYYBOVa1i0TD85RjlEkFjaWDxSFUx3#scrollTo=zOsoqPdXHuL5&line=16&uniqifier=1) is the original where they claim "eps: A tolerance value for numerical stability."

Intuitively, we are taking an expectation over `random_t`, where `random_t = 0` corresponds to the unknown data distribution (in this case the distribution of MNIST) and `random_t = 1` corresponds to our prior (which is known analytically to be jointly Gaussian). My guess is that we only need to learn the transition times `0 < random_t < 1` because the sampling process will take in `random_t = 1` as an input and produce values at `random_t = 0`.

Reply:
this shift and rescale makes t ~ Unif(eps, 1),  maybe there are some problems with t exactly 0

Reply:
perhaps they could have done `t * (1 - 2 * eps) + eps` so that `t' ~ Unif(eps, 1 - eps)`?

Either way you are correct:
``` julia
h ./ expand_dims(unet.layers.marginal_prob_std(t), 3)
```
results in `h ./ 0` for some component when `t=0` since `marginal_prob_std(0) = 0`. So probably just a little fudge factor 😊

Reply:
To me the implementation looks like `random_t` needs to be coming from the data loader

Reply:
Sorry, why can we not construct `random_t` on the fly? They are just randomly sampled indexes into time, they don't exist as a datasource outside the loss function.

Reply:
```suggestion
    random_t = rand!(similar(x, batch_size)) .* (1.0f0 - ϵ) .+ ϵ
```
Generating on the fly seems fine, though you can get away with not using `device`. AFAICT this is the only remaining use in `model_loss`, so the arg can be removed completely.